File: client-domain/src/main/java/com/jraska/github/client/common/Stopwatch.java
package com.jraska.github.client.common;

import com.jraska.github.client.time.RealTimeProvider;
import com.jraska.github.client.time.TimeProvider;

public final class Stopwatch {

  private static final long NOT_SET = -1;

  public static Stopwatch create() {
    return new Stopwatch(RealTimeProvider.INSTANCE);
  }

  public static Stopwatch started() {
    return create().start();
  }

  private final TimeProvider timeProvider;

  private long start = NOT_SET;
  private long stop = NOT_SET;

  Stopwatch(TimeProvider timeProvider) {
    this.timeProvider = Preconditions.argNotNull(timeProvider);
  }

  public Stopwatch start() {
    if (running()) {
      return this;
    }

    if (stop == NOT_SET) {
      start = timeProvider.elapsed();
    } else {
      stop = NOT_SET;
    }

    return this;
  }

  public Stopwatch stop() {
    if (!running()) {
      return this;
    }

    stop = timeProvider.elapsed();
    return this;
  }

  public Stopwatch reset() {
    start = NOT_SET;
    stop = NOT_SET;
    return this;
  }

  public Stopwatch restart() {
    return reset().start();
  }

  public boolean running() {
    return start != NOT_SET && stop == NOT_SET;
  }

  public long time() {
    if (running()) {
      return timeProvider.elapsed() - start;
    } else {
      return stop - start;
    }
  }
}

File: client-domain/src/main/java/com/jraska/github/client/common/Stopwatch.kt
package com.jraska.github.client.common

import com.jraska.github.client.time.RealTimeProvider
import com.jraska.github.client.time.TimeProvider

class Stopwatch(private val timeProvider: TimeProvider) {

  private var start = NOT_SET
  private var stop = NOT_SET

  fun start(): Stopwatch {
    if (running()) {
      return this
    }

    if (stop == NOT_SET) {
      start = timeProvider.elapsed()
    } else {
      stop = NOT_SET
    }

    return this
  }

  fun stop(): Stopwatch {
    if (!running()) {
      return this
    }

    stop = timeProvider.elapsed()
    return this
  }

  fun reset(): Stopwatch {
    start = NOT_SET
    stop = NOT_SET
    return this
  }

  fun restart(): Stopwatch {
    return reset().start()
  }

  fun running(): Boolean {
    return start != NOT_SET && stop == NOT_SET
  }

  fun time(): Long {
    return if (running()) {
      timeProvider.elapsed() - start
    } else {
      stop - start
    }
  }

  companion object {
    private val NOT_SET: Long = -1

    fun create(): Stopwatch {
      return Stopwatch(RealTimeProvider.INSTANCE)
    }

    fun started(): Stopwatch {
      return create().start()
    }
  }
}
